node/network/bridge/src/lib.rs

```diff
@@ -570,31 +592,65 @@ async fn run_network<N: Network>(
 			Action::Abort => return Ok(()),
 
 			Action::SendValidationMessage(peers, msg) => send_message(
-					&mut net,
+					&mut net.network_service,
 					peers,
 					PeerSet::Validation,
 					WireMessage::ProtocolMessage(msg),
 			).await?,
 
 			Action::SendCollationMessage(peers, msg) => send_message(
-					&mut net,
+					&mut net.network_service,
 					peers,
 					PeerSet::Collation,
 					WireMessage::ProtocolMessage(msg),
 			).await?,
 
-			Action::ConnectToValidators(_peer_set, _validators, _res) => {
-				// TODO: https://github.com/paritytech/polkadot/issues/1461
-			}
+			Action::ConnectToAuthorities(peer_set, authorities, res) => {
+				let (priority_group, requests) = match peer_set {
+					PeerSet::Validation => ("parachain_validation".to_owned(), &mut validation_pending_discovery_requests),
+					PeerSet::Collation => ("parachain_collation".to_owned(), &mut collation_pending_discovery_requests),
+				};
+
+				// drop the previous requests
+				// TODO: do we want to handle more than one request at a time?
+				// what would be a cleaning strategy in that case?
+				let _ = requests.clear();
+
+				// collect multiaddress of authorities
+				let mut multiaddresses = HashSet::new();
+				for authority in authorities.clone().into_iter() {
```

Comment:
```suggestion
				for authority in authorities.iter().cloned() {
```

A larger number of smaller allocations, but my understanding is that the compiler is smart enough to reuse the memory from the previous iteration of the loop, which amortizes the cost and reduces the total allocation required.